File: task.py
'''
    Works both for python2 and python3.
    Basic Algorithm:
     1. Read necessary lines as string(char array)
     2. Extract the time in that line as seconds and store it in a list.
     3. Evaluate the lists to get the required results.

    Functions 'extractTimeExtensive' and 'extractTime' have the same purpose.
     However, the former is slower but safer as trailing spaces wouldn't cause an issue.

    Pre-defined function round is being used,
     due to inaccuracy of float (trailing decimal places might not be zero).
'''

def extractTimeExtensive(line, list):
		
    minutes = 0
    milliseconds = 0

    i = 3
    min = True
    while True:
        i += 1
        if line[i] == 's':
            break
        elif line[i] == '.' or line[i] == ' ':
            continue
        elif line[i] == 'm':
            min = False
        elif min:
            minutes *= 10
            minutes += float(line[i])
        else:
            milliseconds *= 10
            milliseconds += float(line[i])

    seconds = (minutes * 60) + (milliseconds / 1000)
    list.append(round(seconds, 3))

def extractTime(line, list):
    seconds = 0
    factor = 0.001
    i = -2
    while True:
        i -= 1
        if line[i] == 'm':
            break
        elif line[i] == '.':
            continue
        else:
            seconds += float(line[i]) * factor
            factor *= 10

    factor = 60
    while True:
        i -= 1
        if line[i] == ' ':
            break
        else:
            seconds += float(line[i]) * factor
            factor *= 10
    list.append(round(seconds, 3))

def analyzeTime(list):
    num = len(list)
    total = 0.0
    mean = 0.0
    variance_numerator = 0.0
    std_dev = 0.0

    for i in range(len(list)):
        total += list[i]

    mean = total / num
    mean = round(mean, 3)

    for i in range(len(list)):
        variance_numerator += (list[i] - mean)**2

    std_dev = (variance_numerator / num)**(0.5)
    std_dev = round(std_dev, 3)

    count = 0
    for i in range(len(list)):
        if (list[i] >= mean - std_dev) and (list[i] <= mean + std_dev):
            count += 1

    return (mean, std_dev, count)

def main():
    real = []
    user = []
    sys = []

    with open('timestat.txt', 'r') as timestat_data:
        for line in timestat_data:
            if line[0] == 'r':
                extractTimeExtensive(line, real)
            elif line[0] == 'u':
                extractTimeExtensive(line, user)
            elif line[0] == 's':
                extractTimeExtensive(line, sys)

    no_of_runs = len(real)

    real = analyzeTime(real)
    user = analyzeTime(user)
    sys = analyzeTime(sys)

    print("Total number of runs: " + str(no_of_runs))
    print("\nAverage Time Statistics:")
    print("real " + str(real[0]) + "s")
    print("user " + str(user[0]) + "s")
    print("sys  " + str(sys[0]) + "s")
    print("\nStandard deviation of Time statistics:")
    print("real " + str(real[1]) + "s")
    print("user " + str(user[1]) + "s")
    print("sys  " + str(sys[1]) + "s")
    print("\nNumber of runs within average - standard deviation to average + standard deviation:")
    print("real " + str(real[2]))
    print("user " + str(user[2]))
    print("sys  " + str(sys[2]))

if __name__ == '__main__':
    main()

File: task.cpp
#include <cmath>
#include <fstream>
#include <iostream>

using namespace std;

/*
  Basic Algorithm:
   1. Read necessary lines as string.
   2. Extract the time in that line as seconds and store it in a integer array.
   3. Evaluate the lists to get the required results.
*/

// This function rounds off upto three decimal places.
// It is being called due to inaccuracy of float (trailing decimal places might not be zero).
float round(float x) {
  float var = (int)(x * 1000 + .5);
  return (float)var / 1000;
}

void extractTime(string line, float *list, int &index) {
  int minutes = 0, milliseconds = 0;
  float seconds = 0;
  bool min = true;

  for (int i = 4;; i++) {
    if (line[i] == 115)
      break;
    else if (line[i] == 109)
      min = false;
    else if ((line[i] > 47) && (line[i] < 58)) {
      if (min) {
        minutes *= 10;
        minutes += line[i] - 48;
      } else {
        milliseconds *= 10;
        milliseconds += line[i] - 48;
      }
    }
  }
  seconds = (minutes * 60) + (float(milliseconds) / 1000);

  list[index] = seconds;
  index++;

  return;
}

int analyzeTime(float *list, int n, float &mean, float &std_dev) {
  float total = 0.0;
  float variance_numerator = 0.0;
  mean = 0.0;
  std_dev = 0.0;

  for (int i = 0; i < n; i++)
    total += list[i];

  mean = float(total) / n;
  mean = round(mean);

  for (int i = 0; i < n; i++)
    variance_numerator += (list[i] - mean) * (list[i] - mean);

  std_dev = sqrt(variance_numerator / n);
  std_dev = round(std_dev);

  int count = 0;
  for (int i = 0; i < n; i++)
    if ((list[i] >= mean - std_dev) && (list[i] <= mean + std_dev))
      count += 1;

  return count;
}

int main() {
  fstream file("timestat.txt", ios::in);

  if (!file.is_open()) {
    cout << "Error\n";
    return -1;
  }

  string line;
  float real[100], user[100], sys[100];
  int real_count = 0, user_count = 0, sys_count = 0;
  while (getline(file, line)) {
    if (line[0] == 114)
      extractTime(line, real, real_count);
    else if (line[0] == 117)
      extractTime(line, user, user_count);
    else if (line[0] == 115)
      extractTime(line, sys, sys_count);
  }

  int no_of_runs = real_count;

  int count[3];
  float mean[3], std_dev[3];

  count[0] = analyzeTime(real, no_of_runs, mean[0], std_dev[0]);
  count[1] = analyzeTime(user, no_of_runs, mean[1], std_dev[1]);
  count[2] = analyzeTime(sys, no_of_runs, mean[2], std_dev[2]);

  cout << "Total number of runs: " << no_of_runs;
  cout << "\n\nAverage Time Statistics:";
  cout << "\nreal " << mean[0] << "s";
  cout << "\nuser " << mean[1] << "s";
  cout << "\nsys  " << mean[2] << "s";
  cout << "\n\nStandard deviation of Time statistics:";
  cout << "\nreal " << std_dev[0] << "s";
  cout << "\nuser " << std_dev[1] << "s";
  cout << "\nsys  " << std_dev[2] << "s";
  cout << "\n\nNumber of runs within average - standard deviation to "
          "average + standard deviation:";
  cout << "\nreal " << count[0];
  cout << "\nuser " << count[1];
  cout << "\nsys  " << count[2];
  cout << endl;

  return 0;
}
